dicomservice/ballerina/dicom_preprocessor.bal

```diff
@@ -0,0 +1,160 @@
+            string diagnostic = string `Supported values for 'Accept' header: ${joinWithComma(dicomweb:MIME_TYPE_JSON,
+                    dicomweb:MIME_TYPE_DICOM_JSON)}`;
+            return dicomweb:createDicomwebError(message, dicomweb:VALIDATION_ERROR,
+                    diagnostic, httpStatusCode = http:STATUS_NOT_ACCEPTABLE);
+        }
+    }
+
+    return requestMimeHeaders;
+}
+
+isolated function createHttpRequestRecord(http:Request request, json|xml? payload) returns HttpRequest & readonly {
+    map<string[]> headers = {};
+    foreach string headerName in request.getHeaderNames() {
+        string[]|http:HeaderNotFoundError headerValues = request.getHeaders(headerName);
+        if headerValues is string[] {
+            headers[headerName] = headerValues;
+        }
+    }
+
+    return {
+        headers: headers.cloneReadOnly(),
+        payload: payload.cloneReadOnly()
+    };
+}
+
+isolated function setDicomContext(DicomContext dicomContext, http:RequestContext httpContext) {
+    httpContext.set(DICOM_CONTEXT_PROP_NAME, dicomContext);
+}
+
+isolated function getNextService(http:RequestContext context) returns http:NextService?|dicomweb:Error {
+    http:NextService|error? next = context.next();
+    if next is error {
+        string message = "Error occurred while retrieving next HTTP service";
+        return dicomweb:createInternalDicomwebError(message, dicomweb:PROCESSING_ERROR);
+    }
+    return next;
+}
+
+isolated function processQueryParams(map<string[]> queryParams, dicomweb:ResourceType resourceType,
+        map<QueryParamConfig> queryParamConfigMap) returns dicomweb:QueryParameterMap|dicomweb:Error {
+    dicomweb:QueryParameterMap processedParams = {};
+
+    // Dicom attribute/value pairs for match query parameter
+    dicomweb:MatchParameterMap matchParameters = {};
+
+    foreach [string, string[]] [param, value] in queryParams.entries() {
+        if queryParamConfigMap.hasKey(param) { // If there's a pre-processor given in the param config
+            QueryParamConfig paramConfig = queryParamConfigMap.get(param);
+            if paramConfig.active {
+                QueryParamPreProcessor? preProcessor = paramConfig.preProcessor;
+                if preProcessor is QueryParamPreProcessor {
+                    dicomweb:QueryParameterValue processedValue = check preProcessor(value);
+                    processedParams[param] = processedValue;
+                }
+            } else {
+                string message = string `Unsupported query parameter: ${paramConfig.name}`;
+                string diagnostic = string `Supported query parameters: ` +
+                        string `${joinWithComma(...extractActiveQueryParameterNames(queryParamConfigMap))}`;
+                return dicomweb:createDicomwebError(message, diagnostic = diagnostic,
+                        httpStatusCode = http:STATUS_NOT_IMPLEMENTED);
+            }
```

Comment:
Best practice is to use early returns when possible rather than going for nested if-else. Check other places as well and apply wherever possible.

https://learn-ballerina.github.io/best_practices/early_returns.html?highlight=nested#use-early-returns

Reply:
Refactored to use early returns where possible in https://github.com/ballerina-platform/module-ballerinax-health.dicom/commit/3807383514d097fe372c936aa46ef84a627a58d3